validator.js

```diff
@@ -0,0 +1,112 @@
+  if (/[^,] /.test(selector)) {
+    // eslint-disable-next-line no-console
+    console.warn('Try to avoid nested CSS selectors in `createRule`: %o', selector)
+  }
+  if (checks) {
+    checks.map((func, i) => {
+      const checkId = `${id}_${i}`
+      if (all) all.push(checkId)
+      generatedChecks.push({
+        id: checkId,
+        evaluate: func,
+        /* eslint-disable-next-line @typescript-eslint/ban-ts-comment */
+        // @ts-ignore FIXME: axe-core does not list this type. Is it part of the API?
+        metadata: {impact: 'critical'}
+      })
+    })
+  }
+  /* eslint-disable-next-line @typescript-eslint/ban-ts-comment */
+  // @ts-ignore FIXME: axe-core does not list metadata in part of its type. Is it part of the API?
+  generatedRules.push({id, excludeHidden, selector, all, any, metadata: {help, helpUrl}})
+}
+
+function checkForInput(htmlString) {
+  let container
+  if (htmlString.ownerDocument) {
+    // axe tests
+    container = htmlString.ownerDocument.getElementById('mocha-fixture')
+  } else {
+    // javascript tests
+    container = document.getElementById('mocha-fixture')
+    container.innerHTML = htmlString
+  }
+
+  return !!container.getElementsByTagName('INPUT').length
+}
+function checkForOptionalClearButton(htmlString) {
+  let container
+  if (htmlString.ownerDocument) {
+    // axe tests
+    container = htmlString.ownerDocument.getElementById('mocha-fixture')
+  } else {
+    // javascript tests
+    container = document.getElementById('mocha-fixture')
+    container.innerHTML = htmlString
+  }
+
+  const [input] = container.getElementsByTagName('INPUT')
+  const clearButtonId = `${input.id || input.getAttribute('name')}-clear`
+  const clearButton = document.getElementById(clearButtonId)
+  if (clearButton && !(clearButton instanceof HTMLButtonElement)) {
+    return false
+  }
+  return true
+}
+createRule({
+  id: 'for-input',
+  // excludeHidden: false,
+  selector: 'auto-complete',
+  help: 'To know about input you should know X',
+  checks: [checkForInput]
+})
+createRule({
+  id: 'for-clear-button',
+  // excludeHidden: false,
+  selector: 'auto-complete',
+  help: 'To know about clear button you should know Y',
+  checks: [checkForOptionalClearButton]
+})
+
+const validateDOMUsage = {
+  /* eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types */
+  jsFunction(htmlString) {
+    const inputCheck = checkForInput(htmlString)
+    if (!inputCheck) {
+      return {
+        isValid: false,
+        errorMessage: 'no input element found'
+      }
+    }
+    const optionalClearButtonCheck = checkForOptionalClearButton(htmlString)
+    if (!optionalClearButtonCheck) {
+      return {
+        isValid: false,
+        errorMessage: 'clear button must be a button element'
+      }
+    }
+    return {isValid: true}
+  },
+  axeChecks: {
+    checks: generatedChecks,
+    rules: generatedRules
+  }
+}
+
+export default validateDOMUsage
```

Comment:
We could be a bit clever here and export things in a more ergonomic way for users of this module.

```suggestion
export function validate(htmlString) {
    const inputCheck = checkForInput(htmlString)
    if (!inputCheck) {
      return {
        isValid: false,
        errorMessage: 'no input element found'
      }
    }
    const optionalClearButtonCheck = checkForOptionalClearButton(htmlString)
    if (!optionalClearButtonCheck) {
      return {
        isValid: false,
        errorMessage: 'clear button must be a button element'
      }
    }
    return {isValid: true}
  }
}

export const checks = generatedChecks
export const rules = generatedRules

export default {
  checks: generatedChecks
  rules: generatedRules
}
```

This way the usage slightly changes:

```js
import autoCompleteRules from '@github/auto-complete-element/validator.js'
axe.configure(autoCompleteRules)
```

or

```js
import {checks, rules} from '@github/auto-complete-element/validator.js'
const otherRules = [ /* other rules */ ]
const otherChecks = [ /* other rules */ ]
axe.configure({
  rules: [...otherRules, rules]
  checks: [...otherChecks, checks]
})
```

or without axe:

```js
import {validate} from '@github/auto-complete-element/validator.js'
validate(document.body.innerHTML)
```

Reply:
I like it, doing!